kahuna/public/js/services/label.js

```diff
@@ -2,18 +2,43 @@ import angular from 'angular';
 
 var labelService = angular.module('kahuna.services.label', []);
 
-labelService.factory('labelService', ['$rootScope', '$q', function ($rootScope, $q) {
+labelService.factory('labelService',
+                     ['$rootScope', '$q', 'apiPoll', 'imageAccessor',
+                      function ($rootScope, $q, apiPoll, imageAccessor) {
+
+    function readLabelName(label) {
+        return label.data;
+    }
+
+    function readLabelsName(labels) {
+        return labels.map(readLabelName);
+    }
+
+    function labelsEquals(labelsA, labelsB) {
+        return angular.equals(readLabelsName(labelsA), readLabelsName(labelsB));
+    }
+
+    function untilLabelsEqual(image, expectedLabels) {
+        return image.get().then(apiImage => {
+            const apiLabels = imageAccessor.readLabels(apiImage);
+            if (labelsEquals(apiLabels, expectedLabels)) {
+                return apiImage;
+            } else {
+                return $q.reject();
+            }
+        });
+    }
+
     function remove (image, label) {
-        var existingLabels = image.data.userMetadata.data.labels.data;
+        var existingLabels = imageAccessor.readLabels(image);
         var labelIndex = existingLabels.findIndex(lbl => lbl.data === label);
-
         if (labelIndex !== -1) {
             return existingLabels[labelIndex]
                 .delete()
-                .then(newLabels => {
-                    image.data.userMetadata.data.labels = newLabels;
-                    $rootScope.$emit('image-updated', image, image);
-                    return image;
+                .then(newLabels => apiPoll(() => untilLabelsEqual(image, newLabels.data)))
```

Comment:
Yes, but the problem is that we want a new instance of the whole image resource to replace it throughout. Otherwise we're just monkey patching the resource object and that means immutable collections holding image resources can't detect the resource has changed (without doing a deep comparison which would be very inefficient).

In the future I'd like to be able to reflect the change on the client-side only, as we used to do, but it should still replace the whole image resource so the change can be detected. Introducing polling is the simplest workaround to do this for now but I agree it's lame.

Does that make sense?